File: Classes/TiGooglemapsPolygonProxy.m
/**
 * Ti.GoogleMaps
 * Copyright (c) 2015-present by Hans Knöchel. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiGooglemapsPolygonProxy.h"

@implementation TiGooglemapsPolygonProxy

- (GMSPolygon *)polygon
{
  if (_polygon == nil) {
    _polygon = [GMSPolygon new];
    _polygon.tappable = YES;
  }

  return _polygon;
}

#pragma mark Public API's

- (void)setPoints:(id)points
{
  ENSURE_UI_THREAD(setPoints, points);
  ENSURE_TYPE_OR_NIL(points, NSArray);

  [self replaceValue:points forKey:@"points" notification:NO];

  GMSMutablePath *path = [GMSMutablePath path];

  if (points != nil) {
    if ([points count] < 2) {
      NSLog(@"[WARN] Ti.GoogleMaps: You need to specify at least 2 points to create a polygon.");
      return _polygon;
    }

    for (id point in points) {
      if ([point isKindOfClass:[NSDictionary class]]) {
        CLLocationDegrees latitude = [TiUtils doubleValue:[point valueForKey:@"latitude"]];
        CLLocationDegrees longitude = [TiUtils doubleValue:[point valueForKey:@"longitude"]];

        [path addLatitude:latitude longitude:longitude];
      } else if ([point isKindOfClass:[NSArray class]]) {
        CLLocationDegrees latitude = [TiUtils doubleValue:[point objectAtIndex:0]];
        CLLocationDegrees longitude = [TiUtils doubleValue:[point objectAtIndex:1]];

        [path addLatitude:latitude longitude:longitude];
      }
    }
  }

  [[self polygon] setPath:path];
}

- (void)setTappable:(id)value
{
  ENSURE_UI_THREAD(setTappable, value);
  ENSURE_TYPE(value, NSNumber);

  [[self polygon] setTappable:[TiUtils boolValue:value]];
  [self replaceValue:value forKey:@"tappable" notification:NO];
}

- (void)setFillColor:(id)value
{
  ENSURE_UI_THREAD(setFillColor, value);

  [[self polygon] setFillColor:[[TiUtils colorValue:value] _color]];
  [self replaceValue:value forKey:@"fillColor" notification:NO];
}

- (void)setStrokeColor:(id)value
{
  ENSURE_UI_THREAD(setStrokeColor, value);

  [[self polygon] setStrokeColor:[[TiUtils colorValue:value] _color]];
  [self replaceValue:value forKey:@"strokeColor" notification:NO];
}

- (void)setStrokeWidth:(id)value
{
  ENSURE_UI_THREAD(setStrokeWidth, value);

  [[self polygon] setStrokeWidth:[TiUtils floatValue:value def:1]];
  [self replaceValue:value forKey:@"strokeWidth" notification:NO];
}

- (void)setGeodesic:(id)value
{
  ENSURE_UI_THREAD(setGeodesic, value);

  [[self polygon] setGeodesic:[TiUtils boolValue:value def:NO]];
  [self replaceValue:value forKey:@"geodesic" notification:NO];
}

- (void)setTitle:(id)value
{
  ENSURE_UI_THREAD(setTitle, value);
  ENSURE_TYPE(value, NSString);

  [[self polygon] setTitle:[TiUtils stringValue:value]];
  [self replaceValue:value forKey:@"title" notification:NO];
}

- (void)setZIndex:(id)value
{
  ENSURE_UI_THREAD(setZIndex, value);
  ENSURE_TYPE(value, NSNumber);
  
  [[self polygon] setZIndex:[TiUtils intValue:value]];
  [self replaceValue:value forKey:@"zIndex" notification:NO];
}

- (void)setHoles:(id)value
{
  ENSURE_UI_THREAD(setHoles, value);
  ENSURE_TYPE(value, NSArray);

  NSMutableArray *holes = [NSMutableArray arrayWithCapacity:[value count]];

  [value enumerateObjectsUsingBlock:^(id item, NSUInteger idx, BOOL *stop) {
    ENSURE_TYPE(item, NSArray);

    GMSMutablePath *path = [[GMSMutablePath alloc] init];

    for (id point in item) {
      if ([point isKindOfClass:[NSDictionary class]]) {
        CLLocationDegrees latitude = [TiUtils doubleValue:[point valueForKey:@"latitude"]];
        CLLocationDegrees longitude = [TiUtils doubleValue:[point valueForKey:@"longitude"]];

        [path insertCoordinate:CLLocationCoordinate2DMake(latitude, longitude) atIndex:idx];
      } else if ([point isKindOfClass:[NSArray class]]) {
        CLLocationDegrees latitude = [TiUtils doubleValue:[point objectAtIndex:0]];
        CLLocationDegrees longitude = [TiUtils doubleValue:[point objectAtIndex:1]];
        
        [path insertCoordinate:CLLocationCoordinate2DMake(latitude, longitude) atIndex:idx];
      }
    }

    [holes addObject:path];
  }];

  [[self polygon] setHoles:holes];
  [self replaceValue:value forKey:@"holes" notification:NO];
}

- (void)setUserData:(id)value
{
  ENSURE_UI_THREAD(setUserData, value);
  ENSURE_TYPE(value, NSDictionary);
  
  [[self polygon] setUserData:value];
  [self replaceValue:value forKey:@"userData" notification:NO];
}

@end

File: titanium-googlemaps-swift/Classes/TiGooglemapsPolygonProxy.swift
/**
 * Axway Titanium
 * Copyright (c) 2018-present by Axway Appcelerator. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

import UIKit
import TitaniumKit
import GoogleMaps

@objc(TiGooglemapsPolygonProxy)
public class TiGooglemapsPolygonProxy: TiProxy {
  private var polygonInternal: GMSPolygon?
  
  private var polygonInstance: GMSPolygon {
    if let polygon = polygonInternal {
      return polygon
    }
    
    let polygon = GMSPolygon()
    polygon.isTappable = true
    polygonInternal = polygon
    return polygon
  }
  
  private func runOnMainThread(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      TiThreadPerformOnMainThread(block, false)
    }
  }
  
  private func append(point: Any, to path: GMSMutablePath) {
    if let dict = point as? [String: Any] {
      let latitude = TiUtils.doubleValue(dict["latitude"])
      let longitude = TiUtils.doubleValue(dict["longitude"])
      path.addLatitude(latitude, longitude: longitude)
    } else if let array = point as? [Any], array.count >= 2 {
      let latitude = TiUtils.doubleValue(array[1])
      let longitude = TiUtils.doubleValue(array[0])
      path.addLatitude(latitude, longitude: longitude)
    }
  }
  
  @objc public func polygon() -> GMSPolygon {
    return polygonInstance
  }
  
  @objc public func setPoints(_ value: [Any]?) {
    runOnMainThread {
      guard let points = value, points.count >= 2 else {
        NSLog("[WARN] Ti.GoogleMaps: You need to specify at least 2 points to create a polygon.")
        return
      }
      
      let path = GMSMutablePath()
      points.forEach { self.append(point: $0, to: path) }
      self.polygonInstance.path = path
      self.replaceValue(points, forKey: "points", notification: false)
    }
  }
  
  @objc public func setTappable(_ value: Any?) {
    runOnMainThread {
      self.polygonInstance.isTappable = TiUtils.boolValue(value)
      self.replaceValue(value, forKey: "tappable", notification: false)
    }
  }
  
  @objc public func setFillColor(_ value: Any?) {
    runOnMainThread {
      self.polygonInstance.fillColor = TiUtils.colorValue(value)?.color
      self.replaceValue(value, forKey: "fillColor", notification: false)
    }
  }
  
  @objc public func setStrokeColor(_ value: Any?) {
    runOnMainThread {
      self.polygonInstance.strokeColor = TiUtils.colorValue(value)?.color
      self.replaceValue(value, forKey: "strokeColor", notification: false)
    }
  }
  
  @objc public func setStrokeWidth(_ value: Any?) {
    runOnMainThread {
      self.polygonInstance.strokeWidth = CGFloat(TiUtils.floatValue(value, def: 1))
      self.replaceValue(value, forKey: "strokeWidth", notification: false)
    }
  }
  
  @objc public func setGeodesic(_ value: Any?) {
    runOnMainThread {
      self.polygonInstance.geodesic = TiUtils.boolValue(value, def: false)
      self.replaceValue(value, forKey: "geodesic", notification: false)
    }
  }
  
  @objc public func setTitle(_ value: Any?) {
    runOnMainThread {
      let title = TiUtils.stringValue(value)
      self.polygonInstance.title = title
      self.replaceValue(title ?? NSNull(), forKey: "title", notification: false)
    }
  }
  
  @objc public func setZIndex(_ value: Any?) {
    runOnMainThread {
      self.polygonInstance.zIndex = Int32(TiUtils.intValue(value))
      self.replaceValue(value, forKey: "zIndex", notification: false)
    }
  }
  
  @objc public func setHoles(_ value: [Any]?) {
    runOnMainThread {
      guard let holesArray = value else { return }
      var holes: [GMSMutablePath] = []
      
      for case let holePoints as [Any] in holesArray {
        let path = GMSMutablePath()
        holePoints.forEach { self.append(point: $0, to: path) }
        holes.append(path)
      }
      
      self.polygonInstance.holes = holes
      self.replaceValue(value, forKey: "holes", notification: false)
    }
  }
  
  @objc public func setUserData(_ value: Any?) {
    runOnMainThread {
      self.polygonInstance.userData = value
      self.replaceValue(value ?? NSNull(), forKey: "userData", notification: false)
    }
  }
}
